storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCacheTest.java

```diff
@@ -0,0 +1,379 @@
+                RemoteLogSegmentState.COPY_SEGMENT_FINISHED);
+
+        // segment 2
+        // 201 - 300
+        // moved to epoch 3 in between
+        // leader epochs (2, 201), (3, 240)
+        Map<Integer, Long> segment2LeaderEpochs = new HashMap<>();
+        segment2LeaderEpochs.put(2, 201L);
+        segment2LeaderEpochs.put(3, 240L);
+        RemoteLogSegmentMetadata segment2Metadata = createSegmentUpdateWithState(cache, segment2LeaderEpochs, 201L, 300L,
+                RemoteLogSegmentState.COPY_SEGMENT_FINISHED);
+
+        // segment 3
+        // 250 - 400
+        // leader epochs (3, 250), (4, 370)
+        Map<Integer, Long> segment3LeaderEpochs = new HashMap<>();
+        segment3LeaderEpochs.put(3, 250L);
+        segment3LeaderEpochs.put(4, 370L);
+        RemoteLogSegmentMetadata segment3Metadata = createSegmentUpdateWithState(cache, segment3LeaderEpochs, 250L, 400L,
+                RemoteLogSegmentState.COPY_SEGMENT_FINISHED);
+
+        //////////////////////////////////////////////////////////////////////////////////////////
+        // Four segments are added with different boundaries and leader epochs.
+        // Search for cache.remoteLogSegmentMetadata(leaderEpoch, offset)  for different
+        // epochs and offsets
+        //////////////////////////////////////////////////////////////////////////////////////////
+
+        HashMap<EpochOffset, RemoteLogSegmentMetadata> expectedEpochOffsetToSegmentMetadata = new HashMap<>();
+        // Existing metadata entries.
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(1, 40), expectedSegment0Metadata);
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(2, 110), segment1Metadata);
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(3, 240), segment2Metadata);
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(3, 250), segment3Metadata);
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(4, 375), segment3Metadata);
+
+        // Non existing metadata entries.
+        // Search for offset 110, epoch 1, and it should not exist.
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(1, 110), null);
+        // Search for non existing offset 401, epoch 4.
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(4, 401), null);
+        // Search for non existing epoch 5.
+        expectedEpochOffsetToSegmentMetadata.put(new EpochOffset(5, 301), null);
+
+        for (Map.Entry<EpochOffset, RemoteLogSegmentMetadata> entry : expectedEpochOffsetToSegmentMetadata.entrySet()) {
+            EpochOffset epochOffset = entry.getKey();
+            Optional<RemoteLogSegmentMetadata> segmentMetadata = cache.remoteLogSegmentMetadata(epochOffset.epoch, epochOffset.offset);
+            RemoteLogSegmentMetadata expectedSegmentMetadata = entry.getValue();
+            log.info("Searching for {} , result: {}, expected: {} ", epochOffset, segmentMetadata,
```

Comment:
Is this logging needed? Ditto below.

Reply:
We may have this as debug level by default. It will be helpful to see for which `EpochOffset` the test is failed.